### ISSUE TITLE
SMTP OAuth Support for Joomla CMS

### ISSUE THREAD

**Issue by anibalsanchez** [joomla/joomla-cms#43403]
### Is your feature request related to a problem? Please describe.

**Google is phasing out SMTP Plain authentication** for Gmail and other services starting in autumn 2024 in favor of **SMTP OAuth**. This change requires third-party apps to use OAuth, enhancing security and discontinuing less secure access methods.

### Describe the solution you'd like

Integrate SMTP OAuth support in Joomla CMS to maintain email functionality for users relying on Gmail. **Add configuration options for SMTP OAuth in the Global Configuration section**, enabling secure connections with Gmail servers via OAuth.

To accommodate this transition and ensure seamless email functionality for Joomla CMS users who rely on Gmail, the necessary configuration options must be added in the Global Configuration section under **Mail > SMTP Methods**.

The proposed solution involves including settings for configuring **SMTP OAuth**, explicitly utilizing the **XOAUTH2 mechanism**.

The new feature will align Joomla CMS with industry email security standards, provide a smooth transition to OAuth authentication, and improve overall security.

### Additional context

- [Google's transition guide: Transition from less secure apps to OAuth](https://support.google.com/a/answer/14114704?sjid=14283871119718270509-EU)
- [Details on OAuth 2.0: OAuth 2.0 Mechanism](https://developers.google.com/gmail/imap/xoauth2-protocol)

**Comment by jeckodevelopment** [comment]
Please keep in mind that this is related not only to gmail.com email accounts but all the corporate/schools/governments which use Google Workspace for the email management. So implementing this feature would be crucial for those use cases.

Thanks :)

**Comment by jjnxpct** [comment]
I think this is very much needed indeed. I have an increasing amount of customers with Google or Microsoft accounts that need OAuth to connect. Atr this time it's still possible, but the user needs to explicitly allow 'plain' user/pw (or  'unsafe methods' or whatever it's called) to still get this to work. But these providers urge us to use OAuth. I think this will be come a bigger issue when ONLY OAuth is allowed in the future.

**Comment by chrisrjrs** [comment]
I agree this functionality is very much needed indeed.  I see that Microsoft is dropping basic authentication support completely by September 2025. From this article: https://techcommunity.microsoft.com/t5/exchange-team-blog/exchange-online-to-retire-basic-auth-for-client-submission-smtp/ba-p/4114750  I would hope to see support well before then.  Does anyone know if a Mailer settings option to authenticate using OAuth is in the Joomla CMS roadmap?

**Comment by jjnxpct** [comment]
What do we do when Microsoft stops supporting the older / basic authentication options we now use and does not support oAuth? Will this mean we just can't use Microsoft SMTP anymore? I already got clients that tell me their MS support is warning about this when te ask them how to connect to their SMTP from Joomla. 

So I do hope this will be picked up by the Joomla project sometime soon... Unfortunately I am not a coder, so I can't help with that.

**Comment by Hackwar** [comment]
@jjnxpct We are very much willing to include something like this, but we aren't a company like M$ or Wordpress and instead work entirely on volunteer base. So you are welcome to volunteer your time and develop this feature. If you are not a coder, as you wrote, you can also hire another developer to do the work for you. Otherwise you would have to wait for someone else to pick this up, which honestly is not guaranteed to happen anytime soon.

**Comment by jjnxpct** [comment]
@Hackwar I know and you are right.  But your comment kinda makes me feel like I did something wrong here...

**Comment by Hackwar** [comment]
No, you didn't do anything wrong, but at the same time I can't give you any answer that will make you happy. We don't have people who will work on stuff guaranteed and in worst case we wont have this feature in time. And while I would be willing to work on this, I don't have (volunteer) time to do this. So if you really need this, your best bet would be to hire someone to work on this. I mean, the benefit is, that you get free maintenance after it has been merged initially. 😉

**Comment by Ruud68** [comment]
From the google website: https://support.google.com/a/answer/14114704?hl=en

-------
**January 2025**

- Access to less secure apps will be turned off for all Google Accounts.
- CalDAV, CardDAV, IMAP, SMTP, and POP will no longer work with legacy passwords (basic authentication).

**Comment by jeckodevelopment** [comment]
Hi, back on this topic to highlight that without this feature, Joomla websites operating on domains who use Google Workspace (any edition) or Microsoft 365 won't be able to send email. This means, no notifications, no newsletters and no any email coming from the website. 
@Hackwar I know that we're short on people, but we've to keep our core features working and still valid.
A CMS which is not able to send notifications and email won't be so successful on the market.

Can we work on a crowdfunding for the initiative?
Can we estimate the effort?
Can we understand what is technically missing?

Thanks!

**Comment by brianteeman** [comment]
Can we not utilise https://github.com/decomplexity/SendOauth2

**Comment by brianteeman** [comment]
phpmailer has builtin support for using 0auth2 with the decomplexity wrapper. There is an example here https://github.com/PHPMailer/PHPMailer/blob/master/examples/sendoauth2.phps so surely we just need to add configuration options to Joomla to support this.

Or am I missing something here that means everyone is running and hiding?

**Comment by anibalsanchez** [comment]
My two cents: OAuth2 has two main authorization workflows:

**Simple Authorization Workflow**: This workflow directly provides the credentials, access token, and secret in the provider's UI. These credentials remain consistent over time, making this workflow relatively easy to support in the current Mail Settings. However, this workflow is not commonly used.

**Two-Legged Authorization Workflow**: This more complex workflow requires client authentication on the provider platform and a callback to the site to generate the access token. Due to the need for user interaction and a callback mechanism, this workflow is more challenging to support within the existing Mail Settings, primarily designed for credentials entry.

For more details on the two-legged authorization workflow, you can refer to the [Google Identity / Authorization / OAuth 2.0 / OAuth 2.0 for Client-side Web Applications](https://developers.google.com/identity/protocols/oauth2/javascript-implicit-flow).

**Access Token Renewal**: When the access token expires, a refresh token must be used to renew and replace the expired access token.

**Comment by brianteeman** [comment]
From what I read they are both supported - it just needs to be implemented

**Comment by jjnxpct** [comment]
From the Google help pages:

_Starting in March 2025, Google Workspace accounts no longer support less secure apps, third-party apps, or devices that ask you to sign in to your Google Account with your username and password. You must use OAuth to let these apps and devices access your account. Third-party email apps that are no longer supported include Microsoft Outlook and the mail app on iOS and MacOS. For detailed instructions and information, visit Transition from less secure apps to OAuth. For the latest dates, visit Google Workspace Updates._

Also I believe the app passwords are not available anymore and OAuth will be the only way to use Google Gmail service for sending emails. For sites that use Google Gmail in the CMS email settings there will be no other alternative then mayb use PHP mail (I thinks that a bad idea...) or a different external email service that still offers 'plain' SMTP connections.

AcyMailing has implemented OAuth for Google and Microsoft so newsletters ka be send using this sending method. I have tested it and it works. But other extensions like Convert Forms and probably others as well use the Joomla e-mail configuration. So websites using those extensions and als Google Gmail will need a solution for their forms as well.

I am not sure about Microsoft. But i think they will have a similar strategy. 

I hope someone in the Joomla project will know what to do.  Sorry I can't be of anymore use then sharing this info and maybe do some testing when the this is needed.

**Comment by jjnxpct** [comment]
Update: It is starting to happen for us on sites that use Google E-mail. Now I have to look at alternatives. I think there will be a lot more sites that will encounter this issue now.

**Comment by mozaffar** [comment]
This problem is resolved in Wordpress by https://wpmailsmtp.com/ . Also there is another plugin for GMail in Joomla (https://www.web357.com/gmail-smtp-connect-joomla-plugin). So there are some success implemented solutions for this.

If you can implement these in Joomla Core it will be very interesting

**Comment by jjnxpct** [comment]
@mozaffar The Web357 Joomla plugin looks promising. Thanks for pointing that out! I will check it out. The documentation even handles the Admin Tool htaccess blocking of the return URL from Google. I encountered that issue when setting up Google OAuth for Acymailing. 

So this plugin could be a solution for website that need to use Google to send emails through Google servers.

I still think the Joomla project should consider adding Oauth methods (At least for Google and Microsoft) to the CMS mail settings. In the mean time for Google the Web357 plugin might do the trick. 

### Workarounds:

For anyone that runs into this issue: You could consider creating a mailbox on a different server (maybe you have your own server or hosting environnement) and use that to authenticate / send emails form you (clients) website(s). And the add a reply to (and maybe a forwarder) to make sure the site managers gets any replies to the send CMS e-mails.

Using PHP Mail could be an way to go, but I think you will run into all kinds of spamfilter issues. You at least have to add your server IP to the SPF record. But I don't think you can add DKIM to the CMS emails so you will have issues with Google and Microsoft blocking your emails to their users.

**Comment by anibalsanchez** [comment]
I opened the issue to avoid the impact on our clients. It was needed a year ago. Today, there is no choice, and third-party plugins are required to implement the latest security practices for the email service.

I'm closing the issue. Thanks.

**Comment by jjnxpct** [comment]
Why is this issue closed? Because there are third party plugins that can do what Joomla can not do? Without considering the option to add this to the Joomla core?  (For MS there is not even no third party solution yet...)

If there are reasons why this should not be in the Joomla core, please inform us about that.  But is this could be a good addition to the core isn't this a valid feature request? Even when maybe there is no one (at this point) to add this to the CMS?

**Comment by anibalsanchez** [comment]
@jjnxpct In a year, the team couldn't develop the support for this feature. Meanwhile, other features and versions were released. It's clear that the priorities were different.

The window of opportunity has closed. Third-party plugins have already fulfilled the requirement and we've already transitioned to a solution.

Tomorrow, if Joomla starts supporting SMTP OAuth, it would be like a joke to the extension developers and administrators that already invested time and resources to migrate.

P.S. Well, it wouldn't be the first time and listen to the same song. ;-)

**Comment by brianteeman** [comment]
thats what happens when people choose personal profit instead of the common good :(

**Comment by jjnxpct** [comment]
I do think the Joomla project dropped the ball here a bit. That's to bad. But I guess these kind of things happen in an Open Source project. Open Source is great, but dependent on volunteers who can not be blamed when the project fails to deliver. It's just what it is.

I don't know if adding this feature - even though late and while there are third party alternatives (partially at this time)  - should just be dismissed. Being able to send e-mails from the CMS using the largest mail providers should be core functionality.

**Comment by bembelimen** [comment]
> I do think the Joomla project dropped the ball here a bit. That's to bad. But I guess these kind of things happen in an Open Source project.

I thought that we all are "The Joomla! Project"? So feel free to step forward to either: 
- develop this feature and contribute it to core 
- pay a developer to create this feature and contribute to core
- collect the money from all interested parties to pay a developer (and see previous post) 

The fact, that it is not in Joomla! core just means it was not important enough that someone did the investment to get it done. If you need help with creating PRs or need support getting a PR properly tested and, when ready, merged, feel free to reach out to me.

**Comment by jjnxpct** [comment]
Unfortunately I am not a developer.

Maybe we could create some kind of crowdfunding for important improvements like this. A Joomla team decides what features are needed in the core and sets a required amount to get that done. This project is placed on a Joomla crowdfunding website. When there is enough money raised for a feature a team of (dedicated / qualified) Joomla developers implements this in the core. This will make it easier for non-developers like me to also contribute to the code development.

**Comment by anibalsanchez** [comment]
@jjnxpct It's a matter of priorities—let's say a problem, related to new security practices, on 10% of active Joomla 4 and 5 websites (120K) isn't a priority for the project, and it's better to develop other features.

**Comment by Hackwar** [comment]
You are welcome to organize that. I don't see that in the Joomla project for political reasons, but you are open to create such a crowdfunding right now and find a developer to code this. Who do you think that "Joomla team" or "Joomla developers" are?

**Comment by bembelimen** [comment]
> Unfortunately I am not a developer.

You don't need to be a developer, you can pay a developer or do the crowdfunding by yourself.

From my experience it's not that easy that e.g. OSM pays certain people because then other high frequent volunteers will be upset why they're not paid.

If there is a need, just gather people and get it paid/done. You're Joomla!, there is no "they".

**Comment by LadySolveig** [comment]
@universewrld 

> Why doesn't joomla open a fundraiser through the https://github.com/opencollective service, where you can also open a fundraiser for certain features for Joomla (like SEO, Mail, etc.)?

Joomla has no employed developers, the teams consist of volunteers. When Joomla collects money now, who gets it?

To summarize:
Get in touch with the community to see if such a feature has a high probability of being implemented in the core.
Organize your own fundraising if you want to see a feature in Joomla. Look out for developers who can implement it and then create a PR together.

What's not helpful:
Pure expectations and service mentality, and snapping at people who make their work available here free of charge and who voluntarily contribute improvements that they probably also need in their everyday work and therefore make available to the community here.

It may be surprising, but I don't think anyone here can live on air and love alone and almost everyone has to earn money with their work.

**Comment by LadySolveig** [comment]
If you want to push SEO improvements, here would have been an opportunity to participate as a sponsor or mentor [Joomla Acaemy](https://github.com/joomla-projects/Joomla-Academy-Onboarding). 
Joomla  is also part at GSOC 2025.

Maybe you should familiarize yourself better with ways to keep up to date. Infos are present on LinkedIn, there is its own Mastodon instance, Joomla User Groups and yes, the community in Mattermost. And certainly many more channels.

**Comment by LadySolveig** [comment]
Personally, I also think that the assumption that Joomla! cannot implement features that have already been taken up by the “market” elsewhere is wrong. of course, the ecosystem and the extension developers should not be harmed. But it is simply wrong to expect that a feature needed by the masses should no longer be implemented just because extension developers were faster. I think it's in everyone's interest that the core should be stable and not overloaded - but at the same time cover the needs of the majority of users.
It's great if third party extension developers can earn their living with it to bridge the gap, but that doesn't mean that it shouldn't be implented in the core as soon as the resources are free.
But that's just my personal opinion as a developer on this topic.

**Comment by LadySolveig** [comment]
If you would like to discuss this topic further, please open a [Discussion](https://github.com/joomla/joomla-cms/discussions) or open a new Issue with a bug report.
I set the last posts to off-topic, as they have nothing more to do with the actual feature request. This makes it unnecessarily difficult for an interested developer to get an overview of the actual feature request.
